R/ExploratorySummarizedExperiment-class.R

```diff
@@ -116,13 +112,64 @@ ExploratorySummarizedExperiment <- function(assays, colData, annotation, idfield
     idfield = idfield, labelfield = labelfield, entrezgenefield = entrezgenefield, assay_measures = assay_measures,
-    contrast_stats = contrast_stats, gene_set_analyses = gene_set_analyses, dexseq_results = dexseq_results, read_reports = read_reports
+    contrast_stats = contrast_stats, gene_set_analyses = gene_set_analyses, dexseq_results = dexseq_results, read_reports = read_reports,
+    gene_set_analyses_tool = gene_set_analyses_tool
   )
 }
+
+check_gene_set_analyses_tool_consistency <- function(gene_set_analyses, gene_set_analyses_tool) {
+  # gene_set_analyses and gene_set_analyses_tool should have the same list of lists
+  # structure. gene_set_analyses_tool should have a single string
+  
+  # Create default structure if necessary:
+  out <- list()
+  
+  if (is.null(gene_set_analyses_tool)) {
+    gene_set_analyses_tool <- list()
+  }
+  for (assay_name in names(gene_set_analyses)) {
+    if (!assay_name %in% names(gene_set_analyses_tool)) {
+      gene_set_analyses_tool[[assay_name]] <- list()
+    }
+    for (gs_type in names(gene_set_analyses[[assay_name]])) {
+      if (! gs_type %in% names(gene_set_analyses_tool[[assay_name]])) {
+        gene_set_analyses_tool[[assay_name]][[gs_type]] <- list()
+      }
+      for (contrast_name in names(gene_set_analyses[[assay_name]][[gs_type]])) {
+        if (! contrast_name %in% names(gene_set_analyses_tool[[assay_name]][[gs_type]])) {
+          gene_set_analyses_tool[[assay_name]][[gs_type]][[contrast_name]] <- "auto"
+        } else {
+          tool_name <- gene_set_analyses_tool[[assay_name]][[gs_type]][[contrast_name]]
+          if (!is.character(tool_name) || length(tool_name) > 1) {
+            stop(paste0("Invalid gene_set_analyses_tool. gene_set_analyses_tool for ",
+                        gs_type, " and ", contrast_name, " should be one of 'auto', 'gsea' or 'roast'. Found ",
+                        paste0(tool_name, collapse=",")))
+          }
+          if (! tool_name %in% c("auto", "gsea", "roast")) {
+            stop(paste0("Invalid gene_set_analyses_tool. gene_set_analyses_tool for ",
+                        gs_type, " and ", contrast_name, " should be one of 'auto', 'gsea' or 'roast'. Found ",
+                        tool_name))
+          }
+        }
+      }
+      # In case gene_set_analyses_tool has other entries, just keep the ones matching gene_set_analyses
+      contrasts_ordered <- names(gene_set_analyses[[assay_name]][[gs_type]])
+      gene_set_analyses_tool[[assay_name]][[gs_type]] <- gene_set_analyses_tool[[assay_name]][[gs_type]][contrasts_ordered]
+    }
+    gene_set_type_names_ordered <- names(gene_set_analyses[[assay_name]])
+    gene_set_analyses_tool[[assay_name]] <- gene_set_analyses_tool[[assay_name]][gene_set_type_names_ordered]
+  }
+  analysis_names_ordered <- names(gene_set_analyses_tool)
+  gene_set_analyses_tool <- gene_set_analyses_tool[analysis_names_ordered]
+  gene_set_analyses_tool
+}
```

Comment:
```suggestion
check_gene_set_analyses_tool_consistency <- function(gene_set_analyses, gene_set_analyses_tool) {
  if (is.null(gene_set_analyses_tool)) {
    gene_set_analyses_tool <- list()
  }
  
  valid_tools <- c("auto", "gsea", "roast")
  
  # Recursive function to mirror structure and validate
  mirror_and_validate <- function(gsa_node, gst_node, path = "") {
    if (is.data.frame(gsa_node)) {
      # Leaf node - should have a tool string
      tool <- if (is.null(gst_node)) "auto" else gst_node
      
      if (!is.character(tool) || length(tool) != 1 || !tool %in% valid_tools) {
        stop(sprintf("Invalid gene_set_analyses_tool at %s: expected one of %s, got %s",
                     path, paste(valid_tools, collapse=", "), paste(tool, collapse=",")))
      }
      return(tool)
    }
    
    # Recurse through structure
    lapply(setNames(names(gsa_node), names(gsa_node)), function(name) {
      mirror_and_validate(gsa_node[[name]], gst_node[[name]], paste0(path, "/", name))
    })
  }
  
  mirror_and_validate(gene_set_analyses, gene_set_analyses_tool)
}
```

I found this function quite hard to understand- my AI friend suggested this, could you try it?